wagtail/wagtailsearch/backends/elasticsearch.py

```diff
@@ -432,40 +432,101 @@ def _get_es_body(self, for_count=False):
         return body
 
     def _do_search(self):
-        # Params for elasticsearch query
-        params = dict(
-            index=self.backend.get_index_for_model(self.query.queryset.model).name,
-            body=self._get_es_body(),
-            _source=False,
-            from_=self.start,
-        )
-
-        params[self.fields_param_name] = 'pk'
+        PAGE_SIZE = 100
 
-        # Add size if set
         if self.stop is not None:
-            params['size'] = self.stop - self.start
+            limit = self.stop - self.start
+        else:
+            limit = None
+
+        use_scroll = limit is None or limit > PAGE_SIZE
+
+        params = {
+            'index': self.backend.get_index_for_model(self.query.queryset.model).name,
+            'body': self._get_es_body(),
+            '_source': False,
+            'from_': self.start,
+            self.fields_param_name: 'pk',
+        }
+
+        if use_scroll:
+            params.update({
+                'scroll': '2m',
+                'size': PAGE_SIZE,
+            })
+
+            # Send to Elasticsearch
+            page = self.backend.es.search(**params)
+
+            while True:
+                if len(page['hits']['hits']) == 0:
+                    break
+
+                # Get pks from results
+                pks = [hit['fields']['pk'][0] for hit in page['hits']['hits']]
+                scores = {str(hit['fields']['pk'][0]): hit['_score'] for hit in page['hits']['hits']}
+
+                # Initialise results dictionary
+                results = dict((str(pk), None) for pk in pks)
+
+                # Find objects in database and add them to dict
+                queryset = self.query.queryset.filter(pk__in=pks)
+                for obj in queryset:
+                    results[str(obj.pk)] = obj
+
+                    if self._score_field:
+                        setattr(obj, self._score_field, scores.get(str(obj.pk)))
+
+                # Yield results in order given by Elasticsearch
+                for pk in pks:
+                    if results[str(pk)]:
+                        yield results[str(pk)]
```

Comment:
Define a variable result = results[str(pk)]. This avoids an extra string conversion & and dictionary lookup in almost all cases.